### PR TITLE
fix: always absolute paths when tarring files to a Docker container

### LINKED ISSUE #673: [Bug]: Files don't work when copying from outside running directory
### Testcontainers version

v0.16.0

### Using the latest Testcontainers version?

Yes

### Host OS

Linux

### Host arch

x86_64 GNU/Linux

### Go version

1.19

### Docker version

```shell
Client:
 Version:           20.10.21
 API version:       1.41
 Go version:        go1.19.2
 Git commit:        baeda1f82a
 Built:             Thu Oct 27 21:30:31 2022
 OS/Arch:           linux/amd64
 Context:           default
 Experimental:      true

Server:
 Engine:
  Version:          20.10.21
  API version:      1.41 (minimum version 1.12)
  Go version:       go1.19.2
  Git commit:       3056208812
  Built:            Thu Oct 27 21:29:34 2022
  OS/Arch:          linux/amd64
  Experimental:     false
 containerd:
  Version:          v1.6.10
  GitCommit:        770bd0108c32f3fb5c73ae1264f7e503fe7b2661.m
 runc:
  Version:          1.1.4
  GitCommit:        
 docker-init:
  Version:          0.19.0
  GitCommit:        de40ad0
```

### Docker info

```shell
Client:
 Context:    default
 Debug Mode: false
 Plugins:
  compose: Docker Compose (Docker Inc., 2.13.0)

Server:
 Containers: 1
  Running: 1
  Paused: 0
  Stopped: 0
 Images: 134
 Server Version: 20.10.21
 Storage Driver: overlay2
  Backing Filesystem: extfs
  Supports d_type: true
  Native Overlay Diff: false
  userxattr: false
 Logging Driver: json-file
 Cgroup Driver: systemd
 Cgroup Version: 2
 Plugins:
  Volume: local
  Network: bridge host ipvlan macvlan null overlay
  Log: awslogs fluentd gcplogs gelf journald json-file local logentries splunk syslog
 Swarm: inactive
 Runtimes: io.containerd.runc.v2 io.containerd.runtime.v1.linux runc
 Default Runtime: runc
 Init Binary: docker-init
 containerd version: 770bd0108c32f3fb5c73ae1264f7e503fe7b2661.m
 runc version: 
 init version: de40ad0
 Security Options:
  seccomp
   Profile: default
  cgroupns
 Kernel Version: 6.0.10-arch2-1
 Operating System: EndeavourOS
 OSType: linux
 Architecture: x86_64
 CPUs: 16
 Total Memory: 31.08GiB
 Name: arjen-dell
 ID: 3CVQ:MB6O:CG3P:4X6U:3NIR:LOV6:NWHO:QAH7:7OYO:MMMN:ZJX3:ZTXE
 Docker Root Dir: /var/lib/docker
 Debug Mode: false
 Registry: https://index.docker.io/v1/
 Labels:
 Experimental: false
 Insecure Registries:
  127.0.0.0/8
 Live Restore Enabled: false
```

### What happened?

When using the Files to copy a directory from my host to the container it doesn't copy the contents of the dir but the complete path.

```
req := testcontainers.ContainerRequest{
		Image: "amacneil/dbmate",
		Files: []testcontainers.ContainerFile{{
			HostFilePath:      "/home/myname/migrations/",
			ContainerFilePath: "/tmp",
			FileMode:          700,
		}},
		WaitingFor: wait.ForExit(),
		Cmd:        []string{"up"},
	}
```

I expect that the contents of `/home/myname/migrations/` is copied into `/tmp` in the container.
Instead it creates `/tmp/home/myname/migrations/`

I also tried using `../../migrations` as a path but this would create a `migrations` dir in the root instead of in the configured `/tmp`

### Relevant log output

_No response_

### Additional information

_No response_

### LINKED ISSUE #763: [Bug]: Cannot locate specified Dockerfile when using FromDockerfile
### Testcontainers version

0.17.0

### Using the latest Testcontainers version?

Yes

### Host OS

Windows

### Host arch

x64

### Go version

1.19.4

### Docker version

```shell
$ docker version
Client:
 Cloud integration: v1.0.29
 Version:           20.10.21
 API version:       1.41
 Go version:        go1.18.7
 Git commit:        baeda1f
 Built:             Tue Oct 25 18:08:16 2022
 OS/Arch:           windows/amd64
 Context:           default
 Experimental:      true

Server: Docker Desktop 4.15.0 (93002)
 Engine:
  Version:          20.10.21
  API version:      1.41 (minimum version 1.12)
  Go version:       go1.18.7
  Git commit:       3056208
  Built:            Tue Oct 25 18:00:19 2022
  OS/Arch:          linux/amd64
  Experimental:     false
 containerd:
  Version:          1.6.10
  GitCommit:        770bd0108c32f3fb5c73ae1264f7e503fe7b2661
 runc:
  Version:          1.1.4
  GitCommit:        v1.1.4-0-g5fd4c4d
 docker-init:
  Version:          0.19.0
  GitCommit:        de40ad0
```

### Docker info

```shell
$ docker info
Client:
 Context:    default
 Debug Mode: false
 Plugins:
  buildx: Docker Buildx (Docker Inc., v0.9.1)
  compose: Docker Compose (Docker Inc., v2.13.0)
  dev: Docker Dev Environments (Docker Inc., v0.0.5)
  extension: Manages Docker extensions (Docker Inc., v0.2.16)
  sbom: View the packaged-based Software Bill Of Materials (SBOM) for an image (Anchore Inc., 0.6.0)
  scan: Docker Scan (Docker Inc., v0.22.0)

Server:
 Containers: 2
  Running: 0
  Paused: 0
  Stopped: 2
 Images: 27
 Server Version: 20.10.21
 Storage Driver: overlay2
  Backing Filesystem: extfs
  Supports d_type: true
  Native Overlay Diff: true
  userxattr: false
 Logging Driver: json-file
 Cgroup Driver: cgroupfs
 Cgroup Version: 1
 Plugins:
  Volume: local
  Network: bridge host ipvlan macvlan null overlay
  Log: awslogs fluentd gcplogs gelf journald json-file local logentries splunk syslog
 Swarm: inactive
 Runtimes: io.containerd.runc.v2 io.containerd.runtime.v1.linux runc
 Default Runtime: runc
 Init Binary: docker-init
 containerd version: 770bd0108c32f3fb5c73ae1264f7e503fe7b2661
 runc version: v1.1.4-0-g5fd4c4d
 init version: de40ad0
 Security Options:
  seccomp
   Profile: default
 Kernel Version: 5.10.16.3-microsoft-standard-WSL2
 Operating System: Docker Desktop
 OSType: linux
 Architecture: x86_64
 CPUs: 8
 Total Memory: 12.35GiB
 Name: docker-desktop
 ID: XT4C:W237:JTIN:MZ6Z:QKJQ:2PM2:VHQJ:E2KN:H4DU:M6LQ:2UBY:THFF
 Docker Root Dir: /var/lib/docker
 Debug Mode: false
 HTTP Proxy: http.docker.internal:3128
 HTTPS Proxy: http.docker.internal:3128
 No Proxy: hubproxy.docker.internal
 Registry: https://index.docker.io/v1/
 Labels:
 Experimental: false
 Insecure Registries:
  hubproxy.docker.internal:5000
  127.0.0.0/8
 Live Restore Enabled: false

WARNING: No blkio throttle.read_bps_device support
WARNING: No blkio throttle.write_bps_device support
WARNING: No blkio throttle.read_iops_device support
WARNING: No blkio throttle.write_iops_device support
```

### What happened?

**Describe the bug**

I'm creating a go-service with testcontainer  by using `FromDockerfile` in a test function, but when I run a test, it could not locate the dockerfile  
I already run the test function ```go func Test_BuildContainerFromDockerfile(t *testing.T)``` in file docker_test.go  (in the testcontainer src) but the result still the same as my test 

**My Test function** 
``` go
package main_test

import (
	"context"
	"go-specs-greet/go_specs_greet"
	"go-specs-greet/specifications"
	"log"
	"os/exec"
	"testing"

	"github.com/alecthomas/assert/v2"
	"github.com/testcontainers/testcontainers-go"
	"github.com/testcontainers/testcontainers-go/wait"
)
func TestGreeterServer(t *testing.T) {

	ctx := context.Background()
	cmd := exec.Command("pwd")
	stdout, err := cmd.Output()
	if err != nil {
		log.Fatalln(err)
	}
	path := string(stdout)
	log.Println(path)
	req := testcontainers.ContainerRequest {
		FromDockerfile: testcontainers.FromDockerfile{
			Context: "./docker",
			Dockerfile: "Dockerfile",
			PrintBuildLog: true,
		},
		ExposedPorts: []string{"8080:8080"},
		WaitingFor: wait.ForHTTP("/").WithPort("8080"),
	}

	log.Println(req.FromDockerfile.Context)
	log.Println(req.FromDockerfile.Dockerfile)

	container, err := testcontainers.GenericContainer(ctx, testcontainers.GenericContainerRequest{
		ContainerRequest: req,
		Started: true,
	})

	assert.NoError(t, err)
	t.Cleanup(func() {
		assert.NoError(t, container.Terminate(ctx))
	})

	driver := go_specs_greet.Driver{BaseURL: "http://localhost:8080"}
	specifications.GreetSpecification(t, driver)
}
```

The output of my test
```
$ go test greeter_server_test.go 
2023/01/17 02:37:01 /d/CODE/<my_proj_path>/Scaling acceptance tests/cmd/httpserver

2023/01/17 02:37:01 ./docker
2023/01/17 02:37:01 Dockerfile
2023/01/17 02:37:01 github.com/testcontainers/testcontainers-go - Connected to docker:
  Server Version: 20.10.21
  API Version: 1.41
  Operating System: Docker Desktop
  Total Memory: 12648 MB
2023/01/17 02:37:01 Starting container id: b015bbb76941 image: docker.io/testcontainers/ryuk:0.3.4
2023/01/17 02:37:02 Waiting for container id b015bbb76941 image: docker.io/testcontainers/ryuk:0.3.4
2023/01/17 02:37:02 Container is ready id: b015bbb76941 image: docker.io/testcontainers/ryuk:0.3.4
&{0xc00008cae0}
Error Happended
--- FAIL: TestGreeterServer (0.80s)
    greeter_server_test.go:43: Did not expect an error but got:
        Error response from daemon: Cannot locate specified Dockerfile: Dockerfile: failed to create container
FAIL
FAIL    command-line-arguments  1.182s
FAIL
```

The test function  ```go func Test_BuildContainerFromDockerfile(t *testing.T)``` in `docker_test.go`
```go
func Test_BuildContainerFromDockerfile(t *testing.T) {
	t.Log("getting context")
	ctx := context.Background()
	t.Log("got context, creating container request")
	req := ContainerRequest{
		FromDockerfile: FromDockerfile{
			Context: "./testresources",
		},
		ExposedPorts: []string{"6379/tcp"},
		WaitingFor:   wait.ForLog("Ready to accept connections"),
	}

	redisC, err := prepareRedisImage(ctx, req, t)
	require.NoError(t, err)
	terminateContainerOnEnd(t, ctx, redisC)
}
```

The output when I run a test function  in `docker_test.go`
```
Running tool: C:\Program Files\Go\bin\go.exe test -timeout 30s -run ^Test_BuildContainerFromDockerfile$ github.com/testcontainers/testcontainers-go

2023/01/17 02:55:17 github.com/testcontainers/testcontainers-go - Connected to docker: 
  Server Version: 20.10.21
  API Version: 1.41
  Operating System: Docker Desktop
  Total Memory: 12648 MB
2023/01/17 02:55:17 Starting container id: ff64537c3741 image: docker.io/testcontainers/ryuk:0.3.4
2023/01/17 02:55:18 Waiting for container id ff64537c3741 image: docker.io/testcontainers/ryuk:0.3.4
2023/01/17 02:55:18 Container is ready id: ff64537c3741 image: docker.io/testcontainers/ryuk:0.3.4
&{0xc0000d48a0}
Error Happended
--- FAIL: Test_BuildContainerFromDockerfile (0.81s)
    c:\Users\<user_name>\go\pkg\mod\github.com\testcontainers\testcontainers-go@v0.17.0\docker_test.go:1066: getting context
    c:\Users\<user_name>\go\pkg\mod\github.com\testcontainers\testcontainers-go@v0.17.0\docker_test.go:1068: got context, creating container request
    c:\Users\<user_name>\go\pkg\mod\github.com\testcontainers\testcontainers-go@v0.17.0\docker_test.go:1201: creating redis container
    c:\Users\<user_name>\go\pkg\mod\github.com\testcontainers\testcontainers-go@v0.17.0\docker_test.go:1205: created redis container
    c:\Users\<user_name>\go\pkg\mod\github.com\testcontainers\testcontainers-go@v0.17.0\docker_test.go:1078: 
        	Error Trace:	c:\Users\<user_name>\go\pkg\mod\github.com\testcontainers\testcontainers-go@v0.17.0\docker_test.go:1078
        	Error:      	Received unexpected error:
        	            	Error response from daemon: Cannot locate specified Dockerfile: Dockerfile: failed to create container
        	Test:       	Test_BuildContainerFromDockerfile
FAIL
FAIL	github.com/testcontainers/testcontainers-go	1.199s
FAIL
```

### Relevant log output

_No response_

### Additional information

This is my project 
[Scaling acceptance tests.zip](https://github.com/testcontainers/testcontainers-go/files/10429133/Scaling.acceptance.tests.zip)

### PR DESCRIPTION
**Pull request by mdelapenya**
- chore: extract to variable to avoid double calculation
- chore: add tests including absolute paths
- chore: merge tests using a test table
- fix: keep parent directory as the root of the TAR file

## What does this PR do?
It modifies how the TAR code creates the TAR file structure, updating the header on each file to remove the absolute path from the tar header. Therefore:

```
/home/myname/migrations
  - a.txr
  - b.txt
  - dir
      - d.txt
```

will generate a tar file from migrations, not keeping the full path before migrations.

Also, as a prevention, we are always converting the source to its absolute representation, which will cover us from wrong interpretations of the Docker libraries when dealing with path on Windows or Linux.

We are adding tests that demonstrates the TAR code behaves as expected in both cases: with absolute paths and with relative ones.

## Why is it important?
Although working at the test level, the users expect that copying files should work as `docker run -v $(pwd):/tmp`, where only contents of the last directory get into the container.

This makes sense to be changed.

## Related issues

- Closes #673
- Closes #763

### CODE REVIEW

File: file.go [hunk 1]
```diff
@@ -57,9 +60,8 @@ func tarDir(src string, fileMode int64) (*bytes.Buffer, error) {
 			return fmt.Errorf("error getting file info header: %w", err)
 		}
 
-		// must provide real name
-		// (see https://golang.org/src/archive/tar/common.go?#L626)
```
Comment (mdelapenya) [inline file.go:61]: Probably worth it to keep this comment? 🤔

File: docker_test.go [hunk 2]
```diff
@@ -2109,11 +2109,23 @@ func TestDockerCreateContainerWithDirs(t *testing.T) {
 	ctx := context.Background()
 	hostDirName := "testresources"
 
+	abs, err := filepath.Abs("./" + hostDirName)
```
Comment (mdelapenya) [inline docker_test.go:2112]:
```suggestion
	abs, err := filepath.Abs(filepath.Join(".", hostDirName))
```

File: file.go [hunk 3]
```diff
@@ -57,9 +60,11 @@ func tarDir(src string, fileMode int64) (*bytes.Buffer, error) {
...
-		header.Name = filepath.ToSlash(file)
+		// see https://pkg.go.dev/archive/tar#FileInfoHeader:
+		// Since fs.FileInfo's Name method only returns the base name of the file it describes,
+		// it may be necessary to modify Header.Name to provide the full path name of the file.
+		// keep the path relative to the parent directory
+		header.Name = filepath.ToSlash(strings.Replace(file, parentDir, "", 1))
```
[the remark this answers is not in the document]
Comment (mdelapenya) [reply file.go:67]: https://pkg.go.dev/path/filepath#ToSlash in theory, package filepath implements utility routines for manipulating filename paths in a way compatible with the target operating system-defined file paths.

[the remark this answers is not in the document]
Comment (HofmeisterAn) [reply file.go:67]: That the path in your tarball, but I refer to this part `strings.Replace(file, parentDir, "", 1)`, like `strings.Replace("C:\Temp\foo\bar\file", "C:/Temp/foo/bar/", "", 1)`.

[the remark this answers is not in the document]
Comment (mdelapenya) [reply file.go:67]: src, and therefore file, comes from the user, so it's their responsibility to pass the proper path. We could sanitise with toSlash the initial src as a mitigation

[the remark this answers is not in the document]
Comment (HofmeisterAn) [reply file.go:67]: But the tests can run on different operating systems.

**Review by kiview — APPROVED** [review]

**Review by HofmeisterAn — APPROVED** [review]

**Pull request merged**